### DISCUSSION TITLE
Adafruit Featherwing OLED 128x64 SH1107

### DISCUSSION THREAD

**Discussion by empirical-dan** [olikraus/u8g2#1461] [Q&A]
Hi
I was just wondering whether I might be able to get some help with this newer OLED display that uses the SH1107 panel. The older 128x32 Featherwing OLED displays are supported but use an SSD1306 panel.
Here is a link to the display including schematics: https://learn.adafruit.com/adafruit-128x64-oled-featherwing?view=all
I have had partial success using the U8X8/U8g2 HelloWorld examples with the following Constructor, having changed the arguments to match my MCU board (Feather 328P - https://learn.adafruit.com/adafruit-feather-328p-atmega328-atmega328p):
U8X8_SH1107_SEEED_128X128_HW_I2C u8x8(/* cs=*/ 7, /* dc=*/ 9, /* reset=*/ 8);
![20210414_160837](https://user-images.githubusercontent.com/52764820/114735187-d7db3c80-9d3c-11eb-9fa3-45f6f782f092.jpg)
Please forgive me, this is my first ever Arduino project and I'm just learning C++ with very limited C for a BSc project in Neuroscience 20 years ago and a little hobby programming as a child in BASIC and Pascal.
I tried changing the dimensions of the display in u8x8_d_sh1107.c (I presume this is part of the driver) to 64x128 or 128x64 but it didn't work.
I also tried rotating the display with U8g2 but it didn't work.
If anyone could point me in the right direction I should eventually be able to work out how to write a driver for this display but I was wondering if it would be relatively straightforward for someone with more knowledge and experience?
Cheers
Dan

**Comment by kmatch98** [comment]
I don’t know much about Arduino but there is a library and guide here. https://learn.adafruit.com/adafruit-128x64-oled-featherwing/arduino-code

Also, you might ask your question on the Adafruit discord server at adafru.it/discord

**Reply by empirical-dan** [reply]
Thanks kmatch.
The board that I'm using (Adafruit feather 328P) has 2k RAM. The adafruit library uses a full page frame buffer only and therefore uses 1k in RAM. It works fine but 1k for my memory allocation is too little (I need just over that and I've refactored my C++ code to minimise memory usage whilst avoiding heap fragmentation). The U8X8 and U8g2 libraries appear to be brilliant if I could get a driver for my panel/controller.

**Reply by olikraus** [reply]
I think it should work. U8x8 may not make sense, because it will appear like shown in your picture. My suggestion is to try U8g2 with R1 option in the SH1107 constructor.

**Comment by empirical-dan** [comment]
Thanks Oli
When I tried previously using R1 in the u8g2 constructor the screen was
blank, presumably because it was rotated off screen. R2 was the same as R0
but the bottom of the screen and upside down. R3 was blank.
Any thoughts how to go about adapting the driver?
Cheers
Dan

**Reply by olikraus** [reply]
oh

**Reply by olikraus** [reply]
I checked this. If you would use the SEEED_128X128 with R1, then probably the visible area of your display starts at 64.
This means you need to print something at location 100 or so to see something. Can you confirm this?

**Comment by empirical-dan** [comment]
Unfortunately the 328P doesn't seem to have enough memory (it only has 2kB) for any of the u8g2 examples (even with page buffer only) it seems so I can't do anything with rotation. With the U8X8 Hello World I can get the beginning of the text to appear on screen by drawing at x = 24.
![20210418_210730](https://user-images.githubusercontent.com/52764820/115159511-59a7ce80-a08b-11eb-8c26-6ee4c659d36a.jpg)
